test/test_fori_loop_with_while_loop_simple_add_dispatch_in_torch.py

```diff
@@ -19,88 +29,311 @@ def _fake_while_loop(cond_fn, body_fn, operands):
+    expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val, out_val))
+    self.assertEqual(res, expected)
+
+  def test_while_loop_tpu_addition_nested(self):
+    xm.mark_step()
+    device = xm.xla_device()
+
+    def cond_fn(iteri, x):
+      return iteri > 0
+
+    def body_fn(iteri, x):
+      return iteri - 1, torch.add(torch.add(x, 1), 1)
+
+    init_val = torch.tensor(2, dtype=torch.int32, device=device)
+    iteri = torch.tensor(10, device=device)
+    _, res =  while_loop(cond_fn, body_fn, (iteri, init_val))
+    _, expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val))
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_simple_linear_inside_loop(self):
+    xm.mark_step()
     device = xm.xla_device()
+    torch.set_grad_enabled(False)
 
-    def cond_fn(init, limit_value):
-      return limit_value[0] >= init[0]
+    n_epochs = 3
+    batch_size_train = 8
+    batch_size_test = 10
+    learning_rate = 0.01
+    momentum = 0.5
+    log_interval = 10
+    random_seed = 1
+    torch.backends.cudnn.enabled = False
+    torch.manual_seed(random_seed)
 
-    def body_fn(init, limit_value):
-      one_value = torch.ones(1, dtype=torch.int32, device=device)
-      return (torch.add(init, one_value), limit_value.clone())
+    ### load data
+    test_loader = xu.SampleGenerator(
+    data=(torch.zeros(8, 1, 28,28), torch.zeros(8, dtype=torch.int64)),
+    sample_count=1000 // 8 // xm.xrt_world_size())
 
-    # TODO(@manfei): init and limit_value has to be torch.tensor.
-    init = torch.tensor([0], dtype=torch.int32, device=device)
-    limit_value = torch.tensor([10], dtype=torch.int32, device=device)
-    res = while_loop(cond_fn, body_fn, (init, limit_value))
-    expected = _fake_while_loop(cond_fn, body_fn, (init, limit_value))
-    self.assertEqual(expected, res)
+    class SimpleLinear(torch.nn.Module):
+      def __init__(self):
+        super().__init__()
+        self.linear = torch.nn.Linear(2, 2)
 
-  def test_while_loop_tpu_subtraction_nested(self):
+      def forward(self, iteri, x):
 
+        def cond_fn(iteri, x):
+          return iteri > 0
+
+        def body_fn(iteri, x):
+          return iteri - 1, self.linear(x)
+
+        return while_loop(cond_fn, body_fn, (iteri, x))
+
+      def forward_compare(self, iteri, x):
+        y = self.linear(x)
+        return iteri - 1, y
+
+    linear_model = SimpleLinear()
+    linear_model.to(device)
+    l_in_0 = torch.randn(2, 2, dtype=torch.float32, device=device)
+    iteri = torch.tensor(2, dtype=torch.int32, device=device)
+    _, res = linear_model(iteri, l_in_0)
+
+    # === expected result after 2 iteration to be compared ===
+    _, expected = linear_model.forward_compare(iteri, l_in_0)
+    _, expected = linear_model.forward_compare(iteri, expected)
+
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_MNIST_inside_loop(self):
+    xm.mark_step()
```

Comment:
ditto, you shouldn't need the `mark_step` here

Reply:
thanks, deleted and updated